feature: boosted creature / reading staticdata from assets / bestiary races rework

Zbizu opened the pull request:
# Description

This PR is quite big, but it would be rather difficult to split these three as they are very closely related to each other.
The original intention was to display boosted creature, but this ended up being a tip of an iceberg - staticdata was not read from the client at all, once I got it working, I found out that data behind boosted creature does not flow properly. Once I fixed that, I found out that the outfits cannot display because the assets don't preload. So in short: a lot of refactoring was involved in order to get a small thing to work.

## Summary of changes
- client is now able to read staticdata (currently it only reads monsters and bosses, but it's easy to add things such as achievements)
- similarly, reading a soundbank should be way easier to implement now
- fixed scrollbar for "hints" panel (huge thanks to **Fifth** for help)
- if service and server are configured in init.lua, the client can now fetch boosted creature/boss the same way QT client does
- **unique server mode only**: in order to display outfits on login screen, the assets are now preloaded
- fixed an issue with wrong characterlist in unique server mode + web login scenario
- RACES and boss equivalent were removed from utils.lua, Cyclopedia outfits are now loaded from staticdata
- corrected the packet for requesting bestiary category
- corrected a few typos in function names and strings

**Removed monster/boss list from cyclopedia utils.lua**
This is a breaking change so I wrote a tool that allows converting staticdata between dat and json:
https://github.com/Zbizu/staticdata-json
There are two branches - `master` repacks staticdata and branch `sound` repacks soundbanks

## Type of change

  - [x] Bug fix (non-breaking change which fixes an issue)
  - [x] New feature (non-breaking change which adds functionality)
  - [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [x] This change requires a documentation update

## How Has This Been Tested

I was testing the code while working on it.
There is some issue with outfits scaling, the one you may know from character list.
Boosted creatures display properly.
Charms assign properly.
Bosses display properly.
Bestiary entries display properly.
Bestiary search bar doesn't seem to do anything. This is due to the fact that the implementation was already incorrect prior to this PR.

**Test Configuration**:

  - Server Version: custom (tfs-based)
  - Client: 13.20
  - Operating System: Windows

## Checklist

  - [x] My code follows the style guidelines of this project
  - [x] I have performed a self-review of my own code
  - [x] I checked the PR checks reports
  - [x] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [x] My changes generate no new warnings
  - [x] I have added tests that prove my fix is effective or that my feature works

Zbizu committed:
correct typos

Zbizu committed:
show exact request type on webservice warning

Zbizu committed:
show warnings on invalid types

Zbizu committed:
add protobufs

Zbizu committed:
fix compilation for multiple proto files

Zbizu committed:
feature: load staticdata

Zbizu committed:
fix characterlist on single server mode

Zbizu committed:
connect to backend

Zbizu committed:
laying groundwork for search feature

Zbizu committed:
cleanup

Zbizu committed:
disable autoload

luanluciano93 commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -496,17 +507,36 @@ function onClickOnNextCalendar()
-            boss_boosted:getCreature():setStaticWalking(1000)
-        end
-    end
+function Boosted_creature(data)
```
```suggestion
function boostedCreature(data)
```

luanluciano93 replied:
the indentation of this entire function is wrong

luanluciano93 commented on modules/client_entergame/entergame.lua:
```diff
@@ -356,74 +362,76 @@ function EnterGame.postEventScheduler()
         end
 
-        modules.client_bottommenu.Booster_creature(response)
+        modules.client_bottommenu.Boosted_creature(response)
```
```suggestion
        modules.client_bottommenu.boostedCreature(response)
```

luanluciano93 commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -496,17 +507,36 @@ function onClickOnNextCalendar()
+		return
+	end
+
+	local raceMon = g_things.getRaceData(data.creatureraceid)
```
```suggestion
	local raceMonster = g_things.getRaceData(data.creatureraceid)
```

luanluciano93 commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -496,17 +507,36 @@ function onClickOnNextCalendar()
+	local fileName = debug.getinfo(1, "S").source -- current file name - bottommenu.lua
+
+	if data.creatureraceid then
+		if raceMon.raceId ~= 0 then
```
```suggestion
		if raceMonster.raceId ~= 0 then
```

luanluciano93 commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -496,17 +507,36 @@ function onClickOnNextCalendar()
+
+	if data.creatureraceid then
+		if raceMon.raceId ~= 0 then
+			monsterOutfit:setOutfit(raceMon.outfit)
```
```suggestion
			monsterOutfit:setOutfit(raceMonster.outfit)
```

luanluciano93 commented on modules/client_entergame/entergame.lua:
```diff
@@ -833,6 +841,13 @@ function EnterGame.setUniqueServer(host, port, protocol, windowWidth, windowHeig
+	-- it needs images of outfits
+    -- so it can display the boosted creature
+	g_game.setClientVersion(clientVersion)
+	g_game.setProtocolVersion(g_game.getClientProtocolVersion(clientVersion))
```
```suggestion
    g_game.setClientVersion(clientVersion)
    g_game.setProtocolVersion(g_game.getClientProtocolVersion(clientVersion))
```

luanluciano93 commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -15,12 +15,21 @@ local eventSchedulerCalendarYearIndex
+
+-- hint 2
+  -- {image = "image of label", Title = "title", description = "your hint here"},
 }
```
```suggestion
    {
        image = "images/randomhint",
	Title = "Enabling Boosted Creature Panel",
	description = "Boosted creatures panel requires configuring a webservice (init.lua) and preloading a client version by either setting one server in Servers_init (init.lua) or by altering entergame.lua.\n\nFor more hints, visit:\t\t https://github.com/mehah/otclient/wiki"
    },

-- hint 2
  -- {image = "image of label", Title = "title", description = "your hint here"},
}
```

Zbizu committed:
resolve variable names

Zbizu committed:
resolve tabbing

Zbizu committed:
debug build fix attempt

Zbizu committed:
Merge branch 'main' into staticdata-rework

dudantas committed:
fix: add /FS flag

dudantas commented on src/client/thingtypemanager.cpp:
```diff
@@ -211,6 +212,89 @@ bool ThingTypeManager::loadAppearances(const std::string& file)
+
+        const auto& raceBank = staticDataLib.monsters();
+        const auto& bossBank = staticDataLib.bosses();
+        //m_monsterRaces.resize(static_cast<size_t>(raceBank.size()) + bossBank.size());
```
You might want to use "reserve" to avoid reallocations and improve performance.

Zbizu replied:
I was close, just forgot the function name😅

dudantas commented on src/client/thingtypemanager.cpp:
```diff
@@ -237,6 +321,28 @@ ThingTypeList ThingTypeManager::findThingTypeByAttr(const ThingAttr attr, const
+            return raceData;
+        }
+    }
+    auto emptyVal = new RaceType();
```
Using new RaceType() introduces unnecessary complexity for managing memory, especially since RaceType is a simple structure. Instead, consider using a static empty instance, such as:

```cpp
static const RaceType emptyVal{};
return emptyVal;
```

Zbizu replied:
is it okay now?

dudantas replied:
> is it okay now?

Yeah. It is good.

dudantas commented on src/client/thingtypemanager.cpp:
```diff
@@ -237,6 +321,28 @@ ThingTypeList ThingTypeManager::findThingTypeByAttr(const ThingAttr attr, const
+
+const std::vector<RaceType>& ThingTypeManager::getRacesByName(const std::string& searchString)
+{
+    static std::vector<RaceType> result;
```
Is it expected for getRacesByName to return the same result across multiple calls? Since result is static, it retains data from previous calls, which could lead to bugs if the function is called multiple times with different input. If that's not the intended behavior, consider using a local variable instead of a static one to ensure the results are independent for each call.

Zbizu replied:
I was trying various ways of pushing this to Lua. Overlooked this one when adapting the code. Thanks!

dudantas committed:
test: move outside if

Zbizu committed:
resolve reviews, reduce code duplication, increase compatibility

Zbizu commented:
changes in fa93c8d:
- addressed some sonar-reported issues
- addressed @dudantas' review
- added compatibility for 'raceid' field which was used before bosstiary was introduced
- added missing 'meta' fields
- reduced code duplication

Zbizu committed:
reuse variables

Zbizu committed:
resolve C4172

dudantas committed:
Revert "test: move outside if"

This reverts commit 601156a96a092b10370aaaa569d4ed81ed71af8f.

dudantas committed:
test: fix compilation

dudantas committed:
try: fix build

dudantas committed:
fix: build

Zbizu commented:
I don't understand why CI has a problem with it
I opened the project as a folder and tried exact same build and it works normally in Windows 10
![obraz](https://github.com/user-attachments/assets/5b5644c8-201e-4bfa-863b-67ae84c88921)

could anyone try with Windows 11?

kokekanon commented:
> I don't understand why CI has a problem with it I opened the project as a folder and tried exact same build and it works normally in Windows 10 ![obraz](https://private-user-images.githubusercontent.com/5205079/405973840-5b5644c8-201e-4bfa-
> could anyone try with Windows 11?


```
>------ Build All started: Project: otclient, Configuration: windows-release-windows-release ------
  [1/31] Running cpp protocol buffer compiler on D:/github/otclient.readme/src/protobuf/staticdata.proto
  [2/31] Running cpp protocol buffer compiler on D:/github/otclient.readme/src/protobuf/sounds.proto
  [3/31] Running cpp protocol buffer compiler on D:/github/otclient.readme/src/protobuf/appearances.proto
  [4/31] Building RC object src\CMakeFiles\otclient.dir\__\cmake\icon\otcicon.rc.res
  [5/31] Building CXX object src\CMakeFiles\otclient.dir\cmake_pch.cxx.obj
  [6/31] Building CXX object src\protobuf\CMakeFiles\protobuf.dir\staticdata.pb.cc.obj
  [7/31] Building CXX object src\protobuf\CMakeFiles\protobuf.dir\sounds.pb.cc.obj
  [8/31] Building CXX object src\protobuf\CMakeFiles\protobuf.dir\appearances.pb.cc.obj
  [9/31] Linking CXX static library src\protobuf\protobuf.lib
  [10/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_20_cxx.cxx.obj
  [11/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_4_cxx.cxx.obj
  [12/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_15_cxx.cxx.obj
  [13/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_14_cxx.cxx.obj
  [14/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_17_cxx.cxx.obj
  [15/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_18_cxx.cxx.obj
  [16/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_19_cxx.cxx.obj
  [17/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_5_cxx.cxx.obj
  [18/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_16_cxx.cxx.obj
  [19/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_13_cxx.cxx.obj
  [20/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_0_cxx.cxx.obj
  [21/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_12_cxx.cxx.obj
  [22/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_7_cxx.cxx.obj
  [23/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_9_cxx.cxx.obj
  [24/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_6_cxx.cxx.obj
  [25/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_11_cxx.cxx.obj
  [26/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_10_cxx.cxx.obj
  [27/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_1_cxx.cxx.obj
  [28/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_3_cxx.cxx.obj
  [29/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_2_cxx.cxx.obj
  [30/31] Building CXX object src\CMakeFiles\otclient.dir\Unity\unity_8_cxx.cxx.obj
  [31/31] Linking CXX executable D:\github\otclient.readme\otclient.exe
  
Build All succeeded.
```

cmake - 
![image](https://github.com/user-attachments/assets/c2e0f611-bcf7-4fc0-8903-837f7025535b)


Edition : Windows 11 Pro 
Version : 23H2 
Os Build 22631.4751 
Installed on 28/‎04/‎2024

Commit:
Merge branch 'main' into staticdata-rework

Zbizu committed:
cmake refactor based on canary

Zbizu committed:
cmake test 2

Zbizu committed:
cmake again

Zbizu commented:
@dudantas @luanluciano93 @mehah build issue fixed

kokekanon reviewed (approved):
test: 
work boosted creature, and cyclopedia  in 13.40

mehah commented:
can i merge then? :)

lucasoares commented on modules/client_bottommenu/bottommenu.lua:
```diff
@@ -496,17 +508,45 @@ function onClickOnNextCalendar()
+
+    -- check if race id is present in the staticdata
+    if raceData.raceId == 0 then
+        local msg = string.format("[%s] Creature with race id %s was not found.", fileName, data.creatureraceid)
```
Have you tested when the raceis not found? This will throw an error since data is not defined. @Zbizu

Zbizu replied:
it returns an empty struct

lucasoares replied:
> it returns an empty struct

For me it was throwing an error saying data is nil...

haha

lucasoares commented:
Have you tested on a 860 server? The bestiary and cyclopedia doesn't exists, but most of servers have the concept of boosted monsters.. I'm unable to make the creature appear in the bottommenu:

```
WARNING: [@/client_bottommenu/bottommenu.lua] Creature with race id 35 was not found.
```

Apparently m_monsterRaces is empty on protocol 860 @Zbizu 

I made it work for simple creatures (with specific look type) by setting outfit with `{type: looktype}` if the race is not found.

Zbizu commented:
@lucasoares 
you need catalog-content.json pointing to staticdata, and a staticdata file of course
you may also need an extra line in things loader for that
would you like to open a pr once you get it working?

Zbizu commented on modules/game_things/things.lua:
```diff
@@ -47,19 +51,17 @@ function load(version)
-            errorMessage = errorMessage ..
-                tr('Unable to load spr file, please place a valid spr in \'%s.spr\'', sprPath)
+            errorList[#errorList + 1] = tr('Unable to load spr file, please place a valid spr in \'%s.spr\'', sprPath)
         end
```
there it needs `g_things.loadStaticData` for things that use spr/dat

you can customize your staticdata with the tool I provided: https://github.com/Zbizu/staticdata-json

lucasoares commented:
> @lucasoares you need catalog-content.json pointing to staticdata, and a staticdata file of course you may also need an extra line in things loader for that would you like to open a pr once you get it working?

Don't think this all is necessary. At least not for the creature boost feature to work. For me it's working now.

I made my status API to return all necessary outfit data:
```
"creatureraceid" => [
                    "type" => $boostedCreature["looktype"],
                    "feet" => $boostedCreature["lookfeet"],
                    "legs" => $boostedCreature["looklegs"],
                    "hand" => $boostedCreature["lookhead"],
                    "body" => $boostedCreature["lookbody"],
                    "addons" => $boostedCreature["lookaddons"]
                ],
```

And made the client compatible with a table response:
```
local function applyToBoostedSlot(data, outfitWidget, imageWidget, fileName)
    -- check if raceId was provided in the JSON response
    if not data then
        return
    end

    local outfit = nil
    if type(data) ~= "table" then
        -- fetch race data
        local raceData = g_things.getRaceData(data)

        -- check if race id is present in the staticdata
        if raceData.raceId == 0 then
            local msg = string.format("[%s] Creature with race id %s was not found.", fileName, data)
            g_logger.warning(msg)

            outfit = {
                type = data,
            }
        else
            outfit = raceData.outfit
        end
    else
        outfit = data
    end

    -- apply to selected widget
    outfitWidget:setOutfit(outfit)
    outfitWidget:getCreature():setStaticWalking(1000)
    outfitWidget:setVisible(true)
    imageWidget:setVisible(false)
end
```

I'm accepting when the race is not found, using the raceid (data) as a looktype (this will work only with looktype-specific creatures, not human outfit, for example) to also have a little more compatibility.
